autogpt_platform/autogpt_libs/autogpt_libs/utils/cache.py

```diff
@@ -138,48 +179,98 @@ async def async_wrapper(*args: P.args, **kwargs: P.kwargs):
-                if key in cache_storage:
-                    if ttl_seconds is None:
-                        logger.debug(
-                            f"Cache hit for {target_func.__name__} args: {args} kwargs: {kwargs}"
+                if shared_cache:
+                    try:
+                        # Use GET directly instead of EXISTS + GET (saves 1 round trip)
+                        cached_bytes = redis_client.get(redis_key)
+                        if cached_bytes is not None and isinstance(cached_bytes, bytes):
+                            logger.debug(
+                                f"Cache hit for {target_func.__name__} args: {args} kwargs: {kwargs}"
+                            )
+                            return pickle.loads(cached_bytes)
+                    except Exception as e:
+                        logger.error(
+                            f"Redis error during cache check for {target_func.__name__}: {e}"
                         )
-                        return cache_storage[key]
-                    else:
-                        cached_data = cache_storage[key]
-                        if isinstance(cached_data, tuple):
-                            result, timestamp = cached_data
-                            if current_time - timestamp < ttl_seconds:
-                                logger.debug(
-                                    f"Cache hit for {target_func.__name__} args: {args} kwargs: {kwargs}"
-                                )
-                                return result
-
-                # Slow path: acquire lock for cache miss/expiry
-                async with _get_cache_lock():
-                    # Double-check: another coroutine might have populated cache
+                        # Fall through to execute function
+                        return await target_func(*args, **kwargs)
+                else:
                     if key in cache_storage:
                         if ttl_seconds is None:
+                            logger.debug(
+                                f"Cache hit for {target_func.__name__} args: {args} kwargs: {kwargs}"
+                            )
                             return cache_storage[key]
                         else:
                             cached_data = cache_storage[key]
                             if isinstance(cached_data, tuple):
                                 result, timestamp = cached_data
                                 if current_time - timestamp < ttl_seconds:
+                                    logger.debug(
+                                        f"Cache hit for {target_func.__name__} args: {args} kwargs: {kwargs}"
+                                    )
                                     return result
 
+                # Slow path: acquire lock for cache miss/expiry
+                async with _get_cache_lock():
+                    # Double-check: another coroutine might have populated cache
+                    if shared_cache:
```

Comment:
The double check here, should it be a lock? Checking a variable might cause issues for a different thread's access